fawltydeps/traverse_project.py

```diff
@@ -22,43 +22,58 @@ def find_sources(
     and yield the corresponding *Source objects found.
     """
 
-    # Collect any directories we will need to traverse.
-    dirs_to_traverse: Dict[Path, Set[Type[Source]]] = {}
+    logger.debug("find_sources() Looking for sources under:")
+    logger.debug(f"    code:   {settings.code}")
+    logger.debug(f"    deps:   {settings.deps}")
+
+    traversal: DirectoryTraversal[Union[Type[Source], Path]] = DirectoryTraversal()
 
     for path_or_special in settings.code if CodeSource in source_types else []:
         # exceptions raised by validate_code_source() are propagated here
         validated: Optional[Source] = validate_code_source(path_or_special)
         if validated is not None:  # parse-able file given directly
+            logger.debug(f"find_sources() Found {validated}")
             yield validated
         else:  # must traverse directory
             # sanity check: convince mypy that SpecialPath is already handled
             assert isinstance(path_or_special, Path)
-            dirs_to_traverse.setdefault(path_or_special, set()).add(CodeSource)
+            traversal.add(path_or_special, CodeSource)
+            traversal.add(path_or_special, path_or_special)  # also record base dir
 
     for path in settings.deps if DepsSource in source_types else []:
         # exceptions raised by validate_deps_source() are propagated here
         validated = validate_deps_source(
             path, settings.deps_parser_choice, filter_by_parser=False
         )
         if validated is not None:  # parse-able file given directly
+            logger.debug(f"find_sources() Found {validated}")
             yield validated
         else:  # must traverse directory
-            dirs_to_traverse.setdefault(path, set()).add(DepsSource)
+            traversal.add(path, DepsSource)
+
+    for _cur_dir, subdirs, files, extras in traversal.traverse():
+        for subdir in subdirs:  # don't recurse into dot dirs
+            if subdir.name.startswith("."):
+                traversal.ignore(subdir)
 
-    for dir_path, types in dirs_to_traverse.items():
+        types = {t for t in extras if t in source_types}
         assert len(types) > 0
-        for file in walk_dir(dir_path):  # traverse directories only _once_
-            if CodeSource in types:
+        if CodeSource in types:
+            # Retrieve base_dir from closest ancestor, i.e. last Path in extras
+            base_dir = next((x for x in reversed(extras) if isinstance(x, Path)), None)
```

Comment:
Double checking my understanding here:
If the user passed 2 code directories, where one is the child of the other, e.g. `bar/` and `bar/foo/`, this line's goal is to select one of those as the `base_dir`. And the way these extras are added as the directory is traversed ensures that the child directory will appear later in the list of extras than the parent directory?

Reply:
Correct. The `extras` list of attached data comes from the `accumulate_attached_data()` helper inside `DirectoryTraversal.traverse()`, and this one guarantees the order of the attached data (parent dirs before child dirs). The line above reverses this, so that - indeed - we do get the "closest" `base_dir`.